Draft PR: KEYCLOAK-18607 add lastUpdatedTimestamp to user

danielFesenmeyer opened the pull request:
This Draft PR adds a lastUpdatedTimestamp to the user (according to KEYCLOAK-18607).

It is currently implemented the following way:
- lastUpdatedTimestamp is set to the same value as the createdTimestamp, when a user is created (Event REGISTER, Admin Event CREATE)
- lastUpdatedTimestamp is updated whenever user profile or email-verified is changed (Events UPDATE_PROFILE and VERIFY_EMAIL, Admin Event UPDATE)
- on import, lastUpdatedTimestamp is handled the same way as createdTimestamp: if specified, it will be copied to the target user, otherwise it will be null
- lastUpdatedTimestamp is NOT updated when:
  - other user-related data is changed like required actions, credentials, consent or federated identities
  - user data (profile) is updated by IDP-/LDAP-Mappers (currently there is no means to figure out if user data actually changed by the mappers and there also is no event for such updates)
- in the Admin Console User Detail View, the lastUpdatedTimestamp is displayed directly below createdTimestamp
 
For our use case, the current implementation would be sufficient, with one exception: We need to get notified when user data is changed by an IDP mapper (and lastUpdatedTimestamp should also be changed in this case).
Currently there is no way to find out whether the UserModel was changed by the IDP Mapper or not. We noticed that MapUserEntity has an isUpdated() method, maybe it makes sense to provide something similar in UserModel?
Or perhaps there is a better way to track the changes?

One additional question: With the current implementation, lastUpdatedTimestamp will be null for users which have been created before the introduction of this timestamp. What do you think about a migration step which sets LAST_UPDATED_TIMESTAMP to CREATED_TIMESTAMP for all users?
Or should this be rather part of a custom migration step?

We appreciate your feedback.

thomasdarimont commented on model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java:
```diff
@@ -105,7 +105,9 @@ public UserModel addUser(RealmModel realm, String id, String username, boolean a
-        entity.setCreatedTimestamp(System.currentTimeMillis());
+        long now = System.currentTimeMillis();
+        entity.setCreatedTimestamp(now);
+        entity.setLastUpdatedTimestamp(now);
```
how about setting the lastUpdate timestamp to null here and really only update it on a real change?

danielFesenmeyer replied:
Having lastUpdated already set on creation has the advantage that it is never null (at least for users created after introduction of this feature).
Furthermore, it has the advantage that readers of a user representation just have to check one timestamp in order to find out when the data was changed/created. (Otherwise they would have to check both created and lastUpdated, which would be less convenient imho.)

thomasdarimont replied:
This makes sense to me and I think your implementation is fine.
However having a dedicated null value would make it easier to check if the user has been updated at all or not, but since this also possible with your change by comparing both timestamps, it should be ok.

Out of couriosity:
Have you looked at how often this update timestamp is updated during normal API interactions?
- Create user?
- Update user directly (enable / disable / change profile field)?
- Update user attribute?
- Change user group / role assignment?

danielFesenmeyer replied:
The current implementation should be aligned with Keycloak Events. In some cases, for example when changing a user via the Admin REST API (Admin Event UPDATE), there might be no "real changes" - but it would be confusing when users receive an UPDATE event, but there is no change to the lastUpdatedTimestamp.

thomasdarimont commented on services/src/main/java/org/keycloak/authentication/requiredactions/ConsoleVerifyEmail.java:
```diff
@@ -92,6 +93,8 @@ public void processAction(RequiredActionContext context) {
             return;
         }
+
+        user.setLastUpdatedTimestampToNow();
```
I'm not sure about the method name... what if we'd need to run additional code on user update?
How about "markAsUpdated()" instead of "setLastUpdatedTimestampToNow"?

danielFesenmeyer replied:
That's a good point. Didn't consider that... I will adjust this.

danielFesenmeyer replied:
@thomasdarimont: I've updated the PR accordingly.

thomasdarimont commented on server-spi/src/main/java/org/keycloak/models/UserModel.java:
```diff
@@ -121,6 +121,22 @@
+     * Marks this user as updated.
+     */
+    default void markAsUpdated() {
+        setLastUpdatedTimestamp(System.currentTimeMillis());
```
I think you should use `Time.currentTimeMillis()` here.

danielFesenmeyer replied:
Thanks for the hint, I adjusted that.

pedroigor reviewed (commented):
@danielFesenmeyer A few general comments:

* Not sure if we need changes to database schema but just leverage user attributes to store the update.
* The User Profile SPI should be the central point for managing user profile aspects. That said, I'm not sure about the `markAsUpdated` method but just a getter to retrieve it from the `UserModel` and the logic to update this data within the `DefaultUserProfile` implementation. That means updating this value only when an attribute changes. Or do you want to also mark as updated for other data (e.g.: required actions) set to the user?

thomasdarimont commented:
> Not sure if we need changes to database schema but just leverage user attributes to store the update.
Is this just to avoid having to migrate large user databases or is there another reason for this? An uppdatedAt timestamp sounds like quite a "core" feature to me that should be provided by the core datamodel, whereas attributes are for user-specific extensions. I know that Keycloak already stores other internal data as user attributes (LDAP references etc.) and I'm not a fan of this, but I wouldn't use attributes for something basic like an updatedAt timestamp.

pedroigor commented:
> Is this just to avoid having to migrate large user databases or is there another reason for this? An updatedAt timestamp sounds like quite a "core" feature to me that should be provided by the core datamodel, whereas attributes are IMHO more suited for user-specific extensions. I know that Keycloak already stores other internal data as user attributes (LDAP references etc. and I'm not a fan of this), but I wouldn't use attributes for something basic like an updatedAt timestamp.

Migration is one concern, although in this case, we are basically updating DDL rather than migrating data (more costly).

We also handle updates to user root attributes (e.g.: email, first, and last names) as attributes, updating their properties accordingly. So I'm wondering if we are moving to store any attribute in the attribute map. We should check with @hmlnarik his thoughts about these changes to storage.

I would say that the decision to have an attribute as a root attribute should take into consideration how it will be used, mainly when querying. So, if we are here more interested in receiving notifications during updates and just using the `updatedAt` when handling events (and when fetching representations), I don't see a reason to change the database schema at all.

For last, LDAP federation also has user attributes for creation/update timestamps. Looks like we need some alignment here too.

danielFesenmeyer commented:
@pedroigor: Thanks for your feedback. I'll try to explain the reasoning behind the current implementation:

> * Not sure if we need changes to database schema but just leverage user attributes to store the update.

We thought that it would be more intuitive to have lastUpdatedTimestamp at the same level as createdTimestamp - which is provided as a "top-level" user property. But storing it as an attribute would be also fine for us.

> * The User Profile SPI should be the central point for managing user profile aspects. That said, I'm not sure about the `markAsUpdated` method but just a getter to retrieve it from the `UserModel` and the logic to update this data within the `DefaultUserProfile` implementation. That means updating this value only when an attribute changes. Or do you want to also mark as updated for other data (e.g.: required actions) set to the user?

From our point of view, the main reason for changing lastUpdatedTimestamp should be changes to the user profile. But there are some other relevant data which are not part of the user profile, but are contained in UserRepresentation (which is included in events and may be relevant to event-listener applications):

- emailVerified: As far as I know, this field is not handled by the DefaultUserProfile implementation. This draft updates lastUpdatedTimestamp when there are changes to this flag.
- disableableCredentialTypes: would be nice if the lastUpdatedTimestamp would change when this property value changes, although this is currently not relevant for us (and **not yet implemented in this draft**)
- requiredActions: as I understand, this rather technical field should be primarily changed by Keycloak itself, so I think that lastUpdatedTimestamp does not need to change in this case
- federatedIdentities: federated identity provider links are relevant for event consumers (at least for our customers), so we would suggest to change lastUpdatedTimestamp when such a link is added or removed (**not yet implemented in this draft**)

Because of this "non-profile" user data, we decided to implement the lastUpdatedTimestamp-handling outside of the DefaultUserProfile implementation.

pedroigor commented:
> We thought that it would be more intuitive to have lastUpdatedTimestamp at the same level as createdTimestamp - which is provided as a "top-level" user property. But storing it as an attribute would be also fine for us.
> 

For me, it also works either way (although using attributes should be much simpler). Let us check with Hynek what he thinks,  considering the upcoming changes to the new store.

> From our point of view, the main reason for changing lastUpdatedTimestamp should be changes to the user profile. But there are some other relevant data which are not part of the user profile, but are contained in UserRepresentation (which is included in events and may be relevant to event-listener applications):

If for you is OK to only take into consideration updates through the user profile, then I would update the PR accordingly. FYI, some initial changes to the email verified flag was done here https://github.com/keycloak/keycloak/pull/8271. I hope we can slowly update similar code to leverage the SPI.

pedroigor commented:
@danielFesenmeyer For now, do not worry about verify email. It was initially addressed by the other PR I mentioned.

For this one, you could remove the `markAsUpdated` method and add to the `DefaultUserProfile` implementation to the necessary logic to set the updatedAt attribute.

This should be enough to make sure this attribute is updated whenever the user is updated through the user profile.

danielFesenmeyer commented:
@pedroigor Ok, thanks for the explanation. Probably it would be good to create a new PR for the "user profile" approach?

pedroigor commented:
@danielFesenmeyer I think it makes more sense to add it here? It should make your changes simpler as you don't need code spread in multiple places for setting the timestamp. It is also more aligned with the idea of the user profile, which you guys also contributed.

pedroigor commented:
@danielFesenmeyer Another approach that would work for dealing with both user attributes and those user properties such as "email verified" is to just change the ISPN `UserAdapter.getDelegateForUpdate` method to set the timestamp whenever that method is called.

However, this does not seem to work for the map storage.

danielFesenmeyer commented:
@pedroigor Your code looks good to me :thumbsup:

There might be a failing test regarding the "email-verify" use case, but this could just be removed if the timestamp is limited to changes to user attributes, as you suggested.
Btw, I see the following advantage in having the timestamp a "top-level" field instead of just an attribute: then there is no logic needed to prevent users from explicitly setting or removing this "special attribute".

pedroigor commented:
@danielFesenmeyer Yeah .. if we are OK with using it for user attributes, then that first set of changes should be OK. Otherwise, I think we can do it within the storage layer, as I mentioned.

danielFesenmeyer commented:
@pedroigor I have updated the PR according to your suggestion to track the changes in the DefaultUserProfile implementation.
I have renamed the timestamp to "attributesUpdatedTimestamp" to make it clear that it's only related to attributes.
(Also rebased to master.)

There are just two points I'm not sure about:
- The current implementation initially set attributesUpdatedTimestamp to the same value as createdTimestamp. This way, for new users those timestamps are the same. But this does not work in the registration use case.
There, technically the user is first created and updated afterwards, leading to a newer attributesUpdatedTimestamp. This is why I couldn't check for equality in the test method _RegisterTest#assertInitialUserTimestamps_. I would say the slightly different timestamp is ok, as long as createdTimestamp is not newer than attributesUpdatedTimestamp. WDYT?
- Caching the information whether attributesUpdatedTimestamp has been already set in KeycloakSession - implemented in _DefaultAttributeChangeListener_. As I understand, this only works when only one user is changed per session - is it safe to make this assumption?

The current implementation still defines the timestamp as a separate field - please let me know if it should be changed to a user attribute.

pedroigor commented:
@danielFesenmeyer The user-profile related changes LGTM. In regards to the new attribute, I'm fine either way, but I would like to see others' opinion.

danielFesenmeyer committed:
KEYCLOAK-18607 add "attributesUpdatedTimestamp" to user

- attributesUpdatedTimestamp is set to the same value as the createdTimestamp, when a user is created
- attributesUpdatedTimestamp is updated whenever a user profile attribute is changed
- on import, attributesUpdatedTimestamp is handled the same way as createdTimestamp: if specified, it will be copied to the target user, otherwise it will be null
- attributesUpdatedTimestamp is NOT updated when other user-related data is changed like required actions, credentials, consent or federated identities
- in the Admin Console User Detail View, attributesUpdatedTimestamp is displayed directly below createdTimestamp

danielFesenmeyer commented:
There is one open issue regarding the IdpReviewProfileAuthentication - I added a comment to the corresponding class. Found this when having a look at test failures. I have an idea how to solve this, but maybe someone comes up with a better idea?

danielFesenmeyer commented:
@pedroigor I have just rebased the code to master. Just some Cross-DC Tests are failing (LifecycleException: Infinispan server startup timed out) - the same as on master.
From my pov the code should be fine now. You mentioned that you would like to see others' opinion. Maybe you can find someone or should I just ask an "arbitrary" maintainer?

pedroigor commented:
@danielFesenmeyer Sorry for the delay. Do we want to continue the discussions around these changes?

danielFesenmeyer commented:
Hi @pedroigor,

Our customers are still interested in a "updatedTimestamp", either on the complete user or just on the user profile attributes.
However, it would be quite important to them, that changes via IDP mappers also lead to a change of that timestamp, and to some kind of "user-update" or "user-profile-update" event.
Only then the timestamp would actually be helpful for them.

What do you think about implementing such functionality?

hmlnarik commented:
This PR touches storage layer in a potentially backwards incompatible way and requires review by the storage team before it can be considered for merging.

pedroigor commented:
@danielFesenmeyer I think we should be able to do it now without touching the core functionality by extending the user profile implementation.

If you write your own, you should be able to add your own custom logic when processing attributes.

Then in the future, we can evaluate incorporating the changes.

danielFesenmeyer commented:
@pedroigor Thanks for your quick response.

Just to make sure that I understood correctly:
You suggest that we implement our own user profile implementation, which stores the timestamp to a certain attribute?
And when it works fine for us, we could file a new PR?

pedroigor commented:
Yes. I don't think we should add this to the server until we have more people asking for it. By writing your own user profile implementation, you can move forward and help us to evaluate later how we can support OOTB.

danielFesenmeyer commented:
@pedroigor: Ok, then I will close this PR for now. Thanks for your explanations. I think it is clear how we can implement a custom user profile implementation. Afaik, my colleague @ioemat has already implemented a similar use case.

einar-lanfranco commented:
@pedroigor  We also need this, actually when we disable a user we cant know the date. Count us in the people asking for!
